FileSystemPath : Add `std::filesystem::path` API

ericmehl opened the pull request:
This adds the ability to set a `Gaffer::FileSystemPath` from a `std::filesystem::path`, and to return a `std::filesystem::path` from a `Gaffer::FileSystemPath`. The same conversion applies to python `pathlib.Path` objects to and from `Gaffer::FileSystemPath` by virtue of the automatic conversion from `pathlib.Path` to `std::filesystem::path`.

I don't know if this is quite the way to go - my reasoning for adding it is :
- There's currently a nasty bug on Windows preventing references from working because of inconsistent path handling. We use `Gaffer.FileSystemPath` there for the filtering capability, and while updating `ReferenceUI` I found myself having to carefully track what was a string vs `Gaffer.FileSystemPath` vs `pathlib.Path`. Keeping everything in `ReferenceUI` as a `pathlib.Path` except for the dialogs seems like an improvement.
- With the ability to hand `pathlib.Path` objects around already, it seems natural to want to be able to give those to a `FileSystemPath` and have it just work.
- There are a few more `as_posix()` I was able to get rid of in the tests.
- We'll eventually migrate `FileSystemPath` to use `std::filesystem::path` instead of `boost::filesystempath`, so hopefully they won't have to coexist for too long?

### Related issues ###

#5003 gives the rationale for keeping `boost::filesystem::path` in `Gaffer::FileSystemPath` for now.

### Breaking changes ###

None

### Checklist ###

- [X] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [X] I have updated the documentation, if applicable.
- [X] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [X] My code follows the Gaffer project's prevailing coding style and conventions.

johnhaddon commented on Changes.md:
```diff
@@ -61,6 +61,9 @@ API
   - ShaderView : `registerReferenceScene()` now takes a `filesystem::path`.
   - Catalogue : `save()` and `load()` take a `filesystem::path`, and `generateFileName()` returns one.
   - ExtensionAlgo : `exportExtension()` now takes a `pathlib.Path` in the `directory` argument.
+  - FileSystemPath :
+    - Added a constructor that accepts a `filesystem::path`.
+    - Added a `path()` method returning a `filesystem::path`.
```
This needs moving to a new version section, since we've released 1.2.0.0a1 now. Maybe 2023 will be the year when I come up with a better way of managing Changes.md.

ericmehl replied:
Fixed in https://github.com/GafferHQ/gaffer/pull/5049/commits/0a73bc6a93e4c09b3c2b7add9edb7eff393c92fb

johnhaddon commented on src/GafferModule/PathBinding.cpp:
```diff
@@ -588,6 +588,13 @@ void GafferModule::bindPath()
 				arg( "includeSequences" ) = false
 			) )
 		)
+		.def(
+			init<const std::filesystem::path &, PathFilterPtr, bool>( (
+				arg( "path" ),
+				arg( "filter" ) = object(),
+				arg( "includeSequences" ) = false
+			) )
+		)
```
Boost evaluates function overloads in reverse order of registration, and takes the first one that it can call with the arguments it has been given. Because `str` can be coerced to `filesystem::path` in our bindings, this means that this overload is taking precedence over the original string version above. We need to move this above L584 to avoid that.

johnhaddon commented on src/Gaffer/FileSystemPath.cpp:
```diff
@@ -252,13 +257,13 @@ bool FileSystemPath::isValid( const IECore::Canceller *canceller ) const
 
-	const file_type t = symlink_status( path( this->string() ) ).type();
-	return t != status_error && t != file_not_found;
+	const boost::filesystem::file_type t = boost::filesystem::symlink_status( boost::filesystem::path( this->string() ) ).type();
+	return t != boost::filesystem::status_error && t != boost::filesystem::file_not_found;
```
Since you need to touch this code anyway, I think we might as well convert to `std::filesystem` here. I believe it's just the time-based stuff in `property()` where we need to stick with `boost` a little longer. If that's right, then I think we can convert most of this file to `std::filesystem`.

ericmehl replied:
I took out all uses except the time related parts in https://github.com/GafferHQ/gaffer/pull/5049/commits/0a73bc6a93e4c09b3c2b7add9edb7eff393c92fb. One step closer to nudging `boost::filesystem` out the airlock.

johnhaddon commented on include/Gaffer/FileSystemPath.h:
```diff
@@ -94,6 +97,8 @@ class GAFFER_API FileSystemPath : public Path
 		// Returns the path converted to the OS native format
 		std::string nativeString() const;
 
+		std::filesystem::path path() const;
```
I wish I could come up with a less ambiguous name for this, since we're mixing `Gaffer::Path` and `filesystem::path` in one API here. Perhaps `stdPath()`? But then that makes for a weird C++-specific name in the Python bindings. Maybe `standardPath()`, where "standard" means "std or pathlib"? I'm not sure. If that doesn't strike you as an improvement, then just stick with what you have...

ericmehl replied:
Yeah, `path()` is not the best wording. I went with `standardPath()` in https://github.com/GafferHQ/gaffer/pull/5049/commits/0a73bc6a93e4c09b3c2b7add9edb7eff393c92fb - it's maybe not perfect but I think it's an improvement over `path()` for clarity.

johnhaddon reviewed (changes requested):
Thanks Eric - this definitely seems like the right way to go.

ericmehl commented:
I addressed the comments and squashed down to https://github.com/GafferHQ/gaffer/pull/5049/commits/0a73bc6a93e4c09b3c2b7add9edb7eff393c92fb. The Windows test is failing, but it looks unrelated. I'm queueing it up again to see if it's maybe a little flaky and a second try will pass.

Other than that test, ready for another look.
